### PR TITLE
refactor(capture,overlay): one jittered retry ladder

### PR DESCRIPTION
**Pull request by tinnhlt-gra**
Wave 6 of the duplication audit, second of three. Item 3 of #2442.

## The copy

`capture/syncstate.go:77` and `overlay/syncbackoff.go:81` carried the same transient-failure ladder **character for character** — the same doubling loop, the same cap clamp, the same `0.8 + 0.4*rand.Float64()`, down to the same `//nolint:gosec` comment with the same reason on the same line of each.

They are sibling modules and cannot import one another, so the copy happened for the honest reason. That is also why it had every opportunity to drift: nothing compared them, and the two are read by different people at different times.

## Why the jitter is the part that had to be shared

The doubling is arithmetic. Anybody would rewrite it the same way, and a mistake in it fails the first test somebody writes.

The ±20% spread is a **decision**, and it is the one that stops a fleet which failed together from retrying together. One provider outage, every connection backing off on the identical schedule, and the recovery arrives as a thundering herd against a provider that has only just come back up. A copy that drifted to ±5%, or dropped the jitter as noise, would look correct in review — and would only ever be visible during an outage, which is the worst possible moment to find out.

So `shared/kernel/backoff.Jittered` owns the shape. Kernel because the callers are two modules and a module never imports a sibling (ADR-0054).

## What deliberately did NOT move

**The numbers.** `base` and `ceiling` bound a ladder against one particular external system, and the two happening to agree on 2min/4h today does not make them one decision — a provider with a tighter budget wants a longer floor, and that is a call somebody makes rather than one kernel makes for them. Each caller keeps its own constants and its own reason for them, and the package's docblock says so where the next author will read it.

This is the same reading as `database.CallerPredicateBudget` in #2677, applied the other way round: there the *number* was one answer to one question asked twice, so the number moved. Here the *shape* is shared and the numbers are genuinely two answers.

**The other ladders.** `webhooks/delivery.go`, `blobstore/s3.go` and `integrations/handoff.go` are different retry policies against different failure modes, as the issue says. I re-measured: they draw no scheduling randomness at all, so they are not this subject and are untouched.

## The gate

`TestTheJitteredLadderIsSpelledOnce` sweeps the negative space — any file outside `kernel/backoff` that draws scheduling randomness fails.

The subject is a call to `math/rand`'s `Float64`, resolved through the **import path** rather than the local name: a file that aliases the package, or one with an unrelated `rand` parameter of its own, would be read wrong by a name match alone. Both `math/rand` and `math/rand/v2` count, because they are two spellings of one package and a gate that knew only the path the tree uses today would go quiet the day it moved.

Mutation-verified: restoring the hand-rolled ladder in `overlay/syncbackoff.go` fails, naming the file. Near-miss fixtures pin the three ways it must stay silent — a `Float64` on something that is not `math/rand`, a file that names `rand` without importing it, and a different `rand` function.

Stated honestly in the header: a jitter built from `IntN`, or a spread computed some other way entirely, is outside the net. It is a net under the one shape the tree reaches for, and the tree holds exactly zero of them outside the owner today.

## The ladder's own tests

Asserted as a **band**, because the jitter is the point — an exact assertion could only pass by removing the thing the package exists for. Four properties, each one a way this can be got wrong:

- the ladder doubles once per prior failure, and `Jittered(0)` waits **one** interval, not two (an off-by-one here doubles every wait in the product);
- it stops at the ceiling, probed at 60 failures — far past where unchecked doubling would overflow an `int64` of nanoseconds and come back negative;
- the jitter actually **varies** (50 calls, ≥10 distinct delays) — a constant multiplier passes every band above while leaving a fleet in lockstep, which is precisely the failure this exists to prevent;
- and it never leaves the band, over 200 draws — a spread that reached zero would retry instantly against a provider that has just refused.

## Verification

`make check-backend` green.

Part of #2442 — the `storekit` cursor envelope is the last of the three.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Unifies capture and overlay transient-failure retries into a single ±20% jittered exponential backoff in `internal/shared/kernel/backoff`, removing duplicate ladders and preventing negative or immediate waits under bad configs or overflow.

- Refactors
  - Adds `backoff.Jittered(priorFailures, base, ceiling)` and switches capture/overlay to use it while keeping their own bounds.
  - Introduces `backend/onejitteredbackoff_test.go` to confine scheduling randomness to `internal/shared/kernel/backoff`, matching draws from `math/rand` and `math/rand/v2`.
  - Adds tests for doubling, first-retry delay, ceiling capping, jitter variance/bounds, and overflow safety.
  - Updates `docs/reference/gate-inventory.md` to register the new gate.

- Bug Fixes
  - Non-positive `base` now returns a 0 delay instead of retrying immediately with a negative duration.
  - Clamps before doubling to prevent overflow that could yield negative waits.
  - Computes the jitter multiply in `float64` and clamps to `MaxInt64` to avoid architecture-dependent overflow.

<sup>Written for commit 4c6350080eab661f6d8cfcd18c08a6f102985f4d. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/margince/margince/pull/2685?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Improvements**
  * Standardized retry delays for capture and overlay synchronization.
  * Retry delays now grow progressively, honor configured limits, and include controlled variation to reduce synchronized retries.
  * Non-positive retry intervals are handled safely without unnecessary delays.

* **Tests**
  * Added comprehensive coverage for retry timing, limits, variation, and edge cases.
  * Added safeguards to keep scheduling randomness centralized.

* **Documentation**
  * Updated the validation inventory with an additional entry.

### CODE REVIEW

**Comment by tinnhlt-gra** [comment]
## Review pass — Codex and Fable

**A base that is not positive was worse than no ladder.** Review asked what `Jittered` does with a zero-value config, and the answer was a defect: doubling zero stays zero, and doubling a **negative** base walks away from the ceiling, so the jitter scales a negative duration and a scheduler asked to wait it comes back **instantly** — against a system that has just failed, which is the one moment it costs. It grows with each failure: five prior failures returned **−25m**.

Both callers pass positive constants, which is exactly why nothing would have caught it. Now they cannot: `Jittered(n, base, ceiling)` returns no wait at all for a base that is not positive, and the test asks both zero and negative across three failure counts. Mutation-verified — removing the guard produces real negative durations.

**One thing Codex flagged that is not true, for the record.** It reported that a zero base makes the loop spin forever. It cannot: the loop is bounded by `priorFailures`, not by reaching the ceiling, so `Jittered(1, 0, ceiling)` terminates immediately. I had repeated that reasoning in my own comment before checking it — the comment now says what actually goes wrong, which is the sign, not the termination.

**Census blindness, declared rather than fixed.** The subject is a call to `math/rand`'s `Float64` resolved through the import path. It cannot see a ladder built from `IntN`, one drawing from a `*rand.Rand` value, or — the one that matters most — a copy that **drops the jitter entirely**, which is the very thing this package exists to keep. A gate for "a ladder without jitter" would have to recognise the ladder rather than the randomness, and I would rather say that plainly than imply a proof the net does not give. The docblock now names all three.

## What review confirmed rather than found

`Jittered` is byte-equivalent to both deleted copies — same loop, same cap, same ±20%, same G404 rationale — and exactly one `rand.Float64` exists in the module. The base/ceiling constants stayed with their callers, each with its own reason, which is the reading opposite to `CallerPredicateBudget` in #2677 and deliberately so: there one number answered one question asked twice, here the shape is shared and the numbers are genuinely two answers.

`make check-backend` green.

File: backend/internal/shared/kernel/backoff/backoff.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	// to wait it retries immediately — against a system that has just failed.
+	if base <= 0 {
+		return 0
+	}
+	delay := base
+	for i := 0; i < priorFailures && delay < ceiling; i++ {
+		delay *= 2
+	}
+	if delay > ceiling {
+		delay = ceiling
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/shared/kernel/backoff/backoff.go:71]: Fixed — clamped before the double rather than after.

Verified at your boundary: `Jittered(1, 1<<62, MaxInt64)` returned **−2248137h59m52s**, and at two and ten prior failures it returned `0s`. A scheduler asked to wait a negative interval retries instantly, against a system that has just failed, which is the one moment this ladder exists for.

Mutation-verified: removing the pre-double clamp reproduces all three.

Worth noting this is the second sign-related defect in the same six lines. The first was a base that is not positive (a negative base doubling *away* from the ceiling); this is a base so large it overflows *past* it. Both end in a negative duration, from opposite ends of the range, and neither is reachable from the two callers — which is exactly why nothing would have caught either.

File: backend/onejitteredbackoff_test.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		if found {
+			return false
+		}
+		call, ok := n.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+		fn, ok := call.Fun.(*ast.SelectorExpr)
+		if !ok || fn.Sel.Name != "Float64" {
+			return true
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/onejitteredbackoff_test.go:101]: Fixed, and you were right that my stated cost was the wrong way round.

I had written "it cannot see a jitter built from `IntN`" into the docblock as a known limitation. That leaves the census able to recognise the **original** while blind to a **copy** — which is backwards for a gate whose entire subject is the copy. A near-miss fixture asserting `IntN` must NOT match was actively holding the hole open, the same way a near miss in the idle-base census was earlier in this sweep.

It now matches the number-drawing functions of both `math/rand` and `math/rand/v2` — `Float64`, `Float32`, `IntN`, `Intn`, `Int63n`, `Int31n`, `Int64N`, `Int32N`, `N` — resolved through the import path so an alias is honoured. Mutation-verified: a ladder in `overlay/syncbackoff.go` spread with `0.8 + 0.004*float64(rand.IntN(100))` is now reported, and was invisible before.

What it still cannot see is stated rather than implied, and one of them matters more than the rest: a draw through a `*rand.Rand` value, a spread computed some other way entirely (a hash of the connection id, a clock's low bits), and **a copy that drops the jitter and keeps only the doubling**. That last would need a detector that recognises the ladder rather than the randomness, and I would rather name it than imply a proof the net does not give.
